src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,467 @@
+/-
+Copyright (c) 2021 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Eric Wieser
+-/
+
+import ring_theory.ideal.basic
+import ring_theory.ideal.operations
+import linear_algebra.finsupp
+import ring_theory.graded_algebra.basic
+/-!
+# Homogeneous ideals of a graded algebra
+
+This file defines homogeneous ideals of `graded_algebra A` where `A : ι → ideal R` and operations on
+them.
+
+## Main definitions
+
+* `ideal.is_homogeneous 𝒜 I`: The property that an ideal is closed under `graded_algebra.proj`.
+* `homogeneous_ideal 𝒜`: The subtype of ideals which satisfy `ideal.is_homogeneous`
+* `ideal.homogeneous_core I 𝒜`: The largest homogeneous ideal smaller than `I`.
+* `ideal.homogeneous_hull I 𝒜`: The smallest homogeneous ideal larger than `I`.
+
+## Main statements
+
+* `homogeneous_ideal.complete_lattice`: `ideal.is_homogeneous` is preserved by `⊥`, `⊤`, `⊔`, `⊓`,
+  `⨆`, `⨅`, and so the subtype of homogeneous ideals inherits a complete lattice structure.
+* `ideal.homogeneous_core.gi`: `ideal.homogeneous_core` forms a galois insertion with coercion.
+* `ideal.homogeneous_hull.gi`: `ideal.homogeneous_hull` forms a galois insertion with coercion.
+
+## Implementation notes
+
+We introduce `ideal.homogeneous_core'` earlier than might be expected so that we can get access
+to `ideal.is_homogeneous.iff_exists` as quickly as possible.
+
+## Tags
+
+graded algebra, homogeneous
+-/
+
+open set_like direct_sum set
+open_locale big_operators pointwise direct_sum
+
+variables {ι R A : Type*}
+
+section homogeneous_def
+
+variables [comm_semiring R] [semiring A] [algebra R A]
+variables (𝒜 : ι → submodule R A)
+variables [decidable_eq ι] [add_monoid ι] [graded_algebra 𝒜]
+variable (I : ideal A)
+
+/--An `I : ideal R` is homogeneous if for every `r ∈ I`, all homogeneous components
+  of `r` are in `I`.-/
+def ideal.is_homogeneous : Prop :=
+∀ (i : ι) ⦃r : A⦄, r ∈ I → (graded_algebra.decompose 𝒜 r i : A) ∈ I
+
+/--For any `comm_ring R`, we collect the homogeneous ideals of `R` into a type.-/
```

Comment:
```suggestion
/-- For any `semiring A`, we collect the homogeneous ideals of `A` into a type. -/
```

Reply:
Actually, generally a lot of the comments refer to `R` but the lemma is about `A`.